core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -112,8 +114,81 @@ class TransactionStateManager(brokerId: Int,
+          transactionMetadataCache.flatMap { case (partition, entry) =>
+            entry.metadataPerTransactionalId.filter { case (_, txnMetadata) => txnMetadata.state match {
+              case Empty | CompleteCommit | CompleteAbort => true
+              case _ => false
+            }
+            }.filter { case (_, txnMetadata) =>
+              txnMetadata.txnLastUpdateTimestamp <= now - config.transactionalIdExpirationMs
+            }.map { case (transactionalId, txnMetadata) =>
+              val txnMetadataTransition = txnMetadata synchronized {
+                txnMetadata.prepareDead
+              }
+              TransactionalIdCoordinatorEpochAndMetadata(transactionalId, entry.coordinatorEpoch, txnMetadataTransition)
+            }
+          }.groupBy { transactionalIdCoordinatorEpochAndMetadata =>
+            partitionFor(transactionalIdCoordinatorEpochAndMetadata.transactionalId)
+          }
+
+        val recordsPerPartition = transactionalIdByPartition
+          .map { case (partition, transactionalIdCoordinatorEpochAndMetadatas) =>
+            val deletes: Array[SimpleRecord] = transactionalIdCoordinatorEpochAndMetadatas.map { entry =>
+              new SimpleRecord(now, TransactionLog.keyToBytes(entry.transactionalId), null)
+            }.toArray
+            val records = MemoryRecords.withRecords(TransactionLog.EnforcedCompressionType, deletes: _*)
+            val topicPartition = new TopicPartition(Topic.TRANSACTION_STATE_TOPIC_NAME, partition)
+            (topicPartition, records)
+          }
+
+
+        def removeFromCacheCallback(responses: collection.Map[TopicPartition, PartitionResponse]): Unit = {
+          responses.foreach { case (topicPartition, response) =>
+            response.error match {
+              case Errors.NONE =>
+                inReadLock(stateLock) {
+                  val toRemove = transactionalIdByPartition(topicPartition.partition())
+                  transactionMetadataCache.get(topicPartition.partition)
+                    .foreach { txnMetadataCacheEntry =>
+                      toRemove.foreach { idCoordinatorEpochAndMetadata =>
+                        val txnMetadata = txnMetadataCacheEntry.metadataPerTransactionalId.get(idCoordinatorEpochAndMetadata.transactionalId)
+                        txnMetadata synchronized {
+                          if (txnMetadataCacheEntry.coordinatorEpoch == idCoordinatorEpochAndMetadata.coordinatorEpoch)
+                            try {
+                              txnMetadata.completeTransitionTo(idCoordinatorEpochAndMetadata.transitMetadata)
+                              txnMetadataCacheEntry.metadataPerTransactionalId.remove(idCoordinatorEpochAndMetadata.transactionalId)
+                            } catch {
+                              case ise: IllegalStateException => debug(s"failed to remove transactionalId: ${idCoordinatorEpochAndMetadata.transactionalId} from cache", ise)
```

Comment:
See comments above; also if we have to call `completeTransitionTo` and it did throw the exception,  should we clear the `pendingState` to None then, otherwise all future requests will be fenced?